File: estc_app/estc_hr/report/employee_leave_and_overtime_summary/employee_leave_and_overtime_summary.js
frappe.query_reports["Employee Leave and Overtime Summary"] = {
    "filters": [
        {
            fieldname: "fiscal_year",
            label: "Academic Year",
            fieldtype: "Link",
            options: "Fiscal Year",
            "reqd": 1,
        },
        {
            fieldname: "department",
            label: "Department",
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                return frappe.db.get_link_options('Department', txt, {"is_group": ['=', 0]});
            },
        },
        {
            fieldname: "leave_type",
            label: "Leave Type",
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                return frappe.db.get_link_options('Leave Type', txt);
            },
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (column.fieldname === 'aluse') {
            value = `<a href="#" class="annual-leave-link" data-employee="${data.employee}" data-employee-name="${data.employee_name}" style="color: blue;">${value}</a>`;
        }
        if (column.fieldname === 'over_time') {
            value = `<a href="#" class="over_time-link" data-employee="${data.employee}" data-employee-name="${data.employee_name}" style="color: blue;">${value}</a>`;
        }
        if (column.fieldname === 'otuse') {
            value = `<a href="#" class="otuse-link" data-employee="${data.employee}" data-employee-name="${data.employee_name}" style="color: blue;">${value}</a>`;
        }
        if (column.fieldname === 'annual_leave') {
            value = `<a href="#" class="annual_leave-link" data-employee="${data.employee}" data-employee-name="${data.employee_name}" style="color: blue;">${value}</a>`;
        }

        return value;
    },

    "onload": function(report) {
        $(document).on("click", ".annual-leave-link", function(e) {
            e.preventDefault();
            var employee = $(this).data('employee');
            var employeeName = $(this).data('employee-name');
            open_report(employee, employeeName , "Annual Leave" , "Leave Use Report");
        });
        $(document).on("click", ".annual_leave-link", function(e) {
            e.preventDefault();
            var employee = $(this).data('employee');
            var employeeName = $(this).data('employee-name');
            open_report(employee, employeeName , "All" , "Leave Use Report");
        });
        $(document).on("click", ".otuse-link", function(e) {
            e.preventDefault();
            var employee = $(this).data('employee');
            var employeeName = $(this).data('employee-name');
            open_report(employee, employeeName , "Over Time" , "Leave Use Report");
        });
        $(document).on("click", ".over_time-link", function(e) {
            e.preventDefault();
            var employee = $(this).data('employee');
            var employeeName = $(this).data('employee-name');
            open_report(employee, employeeName , "" , "OT Report");
        });
    }
};

function open_report(employee, employeeName, leave_type = null, report_name) {
    const fiscalYearFilter = frappe.query_report.get_filter_value("fiscal_year");
    const filters = {
        employee: employee,
        fiscal_year: fiscalYearFilter
    };

    if (leave_type && leave_type != "" ) {
            filters.leave_type = leave_type;
    }

    frappe.call({
        method: "frappe.desk.query_report.run",
        args: {
            report_name: report_name,
            filters: filters
        },
        callback: function(r) {
            if (r.message) {
                var dialog = new frappe.ui.Dialog({
                    title: report_name + " - " + employeeName,
                    size: 'large',
                    fields: [
                        {
                            fieldtype: "HTML",
                            fieldname: "report_content"
                        }
                    ],
                    primary_action_label: "Close",
                    primary_action: function() {
                        dialog.hide();
                    }
                });

                dialog.$wrapper.find('.modal-dialog').css({
                    'max-width': '90%',
                    'width': '90%'
                });

                let content = "";
                let content_ot = "";

                function processAndDisplayContent() {
                    // Combine both tables before rendering
                    let finalContent = content_ot + "<br>" + content;
                    dialog.fields_dict.report_content.$wrapper.html(finalContent);
                    dialog.show();
                }

                    frappe.call({
                        method: "frappe.desk.query_report.run",
                        args: {
                            report_name: "Carry Over Report",
                            filters: {
                                employee: employee,
                                fiscal_year: fiscalYearFilter
                            }
                        },
                        callback: function(response) {
                            if (response.message.result.length > 0) {
                                content_ot = '<h5> Carry Over </h5> <table style="margin-top:0 !important" class="table table-bordered"><thead><tr>';
                                const firstRow = response.message.result[0];
                                Object.keys(firstRow).forEach(key => {
                                    content_ot += `<th>${key}</th>`;
                                });
                                content_ot += '</tr></thead><tbody>';
                                response.message.result.forEach(row => {
                                    content_ot += '<tr>';
                                    Object.values(row).forEach(cell => {
                                        content_ot += `<td>${cell !== null ? cell : ''}</td>`;
                                    });
                                    content_ot += '</tr>';
                                });
                                content_ot += '</tbody></table>';
                            } else {
                                content_ot = "<center>No Data for Carry Over Report</center>";
                            }
                            processAndDisplayContent();
                        }
                    });
                if (typeof r.message === "string") {
                    content = r.message;
                } else if (Array.isArray(r.message.result) && r.message.result.length > 0) {
                    content = '<h5> Leave Use  </h5> <table style="margin-top:0 !important" class="table table-bordered"><thead><tr>';
                    const firstRow = r.message.result[0];
                    Object.keys(firstRow).forEach(key => {
                        content += `<th>${key}</th>`;
                    });
                    content += '</tr></thead><tbody>';
                    r.message.result.forEach(row => {
                        content += '<tr>';
                        Object.values(row).forEach(cell => {
                            content += `<td>${cell !== null ? cell : ''}</td>`;
                        });
                        content += '</tr>';
                    });
                    content += '</tbody></table>';
                } else {
                    content = "<center>No Data</center>";
                }

                if (report_name !== "Leave Use Report") {
                    processAndDisplayContent();
                }
            } else {
                console.log("No data returned for the report.");
            }
        },
        error: function(error) {
            console.error("Error fetching report:", error);
        }
    });
}

File: estc_app/estc_hr/report/employee_leave_and_overtime_summary/employee_leave_and_overtime_summary.py
import frappe

def execute(filters=None):
    return get_columns(filters), get_data(filters)

def get_columns(filters):
    columns = []
    columns.append({'fieldname':"employee",'label':"Employee",'fieldtype':'Link','options':'Employee','align':'center','width':130})
    columns.append({'fieldname':"employee_name",'label':"Employee Name",'align':'left','width':130})
    columns.append({'fieldname':"annual_leave",'label':"Annual Leave",'fieldtype':'Float',"precision":2,'align':'center','width':130, 'click': 'open_annual_leave_report'})
    columns.append({'fieldname':"aluse",'label':"AL USE",'align':'center',"precision":2,'fieldtype':'Float','width':130})
    columns.append({'fieldname':"al_balance",'label':"AL Balance","precision":2,'align':'center','fieldtype':'Float','width':130})
    columns.append({'fieldname':"over_time",'label':"Over Time","precision":2,'align':'center','fieldtype':'Float','width':130})
    columns.append({'fieldname':"otuse",'label':"OT USE","precision":2,'align':'center','fieldtype':'Float','width':130})
    columns.append({'fieldname':"ot_balance",'label':"Ot Balance","precision":2,'align':'center','fieldtype':'Float','width':130})
    columns.append({'fieldname':"alandotbalance",'label':"AL & OT Balance","precision":2,'align':'center','fieldtype':'Float','width':130})
    return columns

def get_data(filters):
    data = []
    
    # Get aggregated leave data for each employee (sum the leave data)
    employees = frappe.db.sql(""" 
        SELECT employee, employee_name,
               SUM(CASE WHEN leave_type = 'Annual Leave' THEN max_leave ELSE 0 END) AS annual_leave,
               SUM(CASE WHEN leave_type = 'Annual Leave' THEN use_leave ELSE 0 END) AS aluse,
               SUM(CASE WHEN leave_type = 'Annual Leave' THEN balance ELSE 0 END) AS al_balance,
               SUM(CASE WHEN leave_type = 'Over Time' THEN max_leave ELSE 0 END) AS over_time,
               SUM(CASE WHEN leave_type = 'Over Time' THEN use_leave ELSE 0 END) AS otuse,
               SUM(CASE WHEN leave_type = 'Over Time' THEN balance ELSE 0 END) AS ot_balance,
               SUM(CASE WHEN leave_type = 'Annual Leave' THEN balance ELSE 0 END) + 
               SUM(CASE WHEN leave_type = 'Over Time' THEN balance ELSE 0 END) AS alandotbalance 
        FROM `tabEmployee Attendance Leave Count`
        {0}
        GROUP BY employee, employee_name
    """.format(get_conditions(filters)), filters, as_dict=True)

    for emp in employees:
        data.append({
            'employee': emp['employee'],
            'employee_name': emp['employee_name'],
            'over_time': emp['over_time'],
            'annual_leave': emp['annual_leave'],
            'aluse': emp['aluse'],
            'al_balance': emp['al_balance'],
            'otuse': emp['otuse'],
            'ot_balance': emp['ot_balance'],
            'alandotbalance': emp['alandotbalance']
        })

    return data

def get_conditions(filters):
    select_filters = []
    
    if filters.get('fiscal_year'):
        select_filters.append("fiscal_year = %(fiscal_year)s")
    
    if filters.get('department') and len(filters['department']) > 0:
        select_filters.append("department in %(department)s")
    
    if filters.get('leave_type') and len(filters['leave_type']) > 0:
        select_filters.append("leave_type in %(leave_type)s")
    
    if len(select_filters) > 0:
        return " WHERE " + " AND ".join(select_filters)
    else:
        return ''
